### ISSUE TITLE
[Dialog] event for single-click in list?

### ISSUE THREAD

**Issue by IgnorantGuru** [IgnorantGuru/spacefm#572]
*Received [elsewhere](https://igurublog.wordpress.com/2015/08/26/spacefm-1-0-3-released/#comment-35078):*
> One of the aspects that most limits me in build my GUIs is, when I use --list, to not have immediate notification of the event of a ‘single click’ on an item in the list .

**Comment by IgnorantGuru** [comment]
@paoloschi A new [--click element](http://ignorantguru.github.io/spacefm/spacefm-manual-next-en.html#dialog-nonvis-click) has been added which should allow you to setup a command for this.  This is now available in the 'next' branch (use [net installer](http://ignorantguru.github.io/spacefm/#installer) to test), due for 1.0.4 release.  Thanks for testing - you can report any issues with this below.

**Comment by IgnorantGuru** [comment]
@paoloschi Also, unrelated to this issue, you may also want to note/test the new [activate](http://ignorantguru.github.io/spacefm/spacefm-manual-next-en.html#sockets-methods-act) method, and the new [evt_tab_chdir](http://ignorantguru.github.io/spacefm/spacefm-manual-next-en.html#sockets-events-tabchdir) event, also in the next branch.

**Comment by paoloschi** [comment]
SpaceFM 1.0.3+next installed with 'spacefm-installer' default values

very very nice implementation: triggering the single-click from other elements in addition to the lists will certainly be a useful thing!
(let me say: I'm walking to 10cm from the floor because of my enjoyment..:)

I started to try using the example in the guide and confirm that it works well here too.
I have a question: how is expected in --list to be the double-click managed in conjunction with the single-click?

case 1: __user, after a while, hits a double-click over the item already selected in list__
- a single event occurs and we have a single notification to manage..

case 2: __user hits a double-click over a item that is not currently selected__
- this should still be considered a single event (!?)  
  the consequence is that 'dialog' reports two simultaneous events, both for the single and double-click

I'm trying to figure, in the second use case, how in my script I have to act then to override the single-click event..
perhaps that the single click could be notified with a some delay? (as the television remote control waits for the eventual selection of a program number with more than one digit?)


another thing that I need to understand..
I assume I need to handle the event by parsing of '%v' in my script, perhaps something like..

    function click_event {
        case ${1%% *} in
          label1) echo "label value = ${1#* }";;
                  # code for 'label' event follows here..

           list1) echo "list value = ${1#* }";;
                  # code for 'list' event follows here..

          input1) echo "input value = ${1#* }"
                  # code for 'input' event follows here..
        esac
      }; export -f click_event

    spacefm -g --label "Click HERE" \
               --list 111 222 333 -- \
               --input \
               --click bash -c 'click_event "%v" >&2'

but I suspect there may be more elegant solutions.. surely there are

and also, guide says: "Multiple --click elements can be added to a dialog."
if I put two --click elements, I see both simultaneously report the same value so what's the point to insert more than one?

..thanks for any clarification to how properly use the new feature.



on the sidelines: I realized that the placeholder `%v` for --click element is replaceable by any name; to say..

    spacefm -g --label "Click HERE" \
               --list 111 222 333 -- bash -c 'echo %foo >&2' \
               --input \
               --click bash -c 'echo %foo >&2'

predictably, prints a blank line for the double-click but continues to print the event for the single-click..
is this to be ignored?..

**Comment by IgnorantGuru** [comment]
Nice job, you found a pre-existing bug with the %foo.  Even in older versions this bug affected elements like --check.  You can see a test for the bug at the top [here](https://github.com/IgnorantGuru/spacefm/commit/5bcdd24b77ed4635c17cf4732a141de4fe6a1527).  This should now be corrected.

> this should still be considered a single event (!?)
> the consequence is that 'dialog' reports two simultaneous events, both for the single and double-click

No, it is not a single event.  The first click focuses the list and changes the selection, which runs the --click command.  The second click of the double-click triggers activation, which runs the --list command (or clicks an ok button if present).  That is normal in general handling of these kinds of events.  For example, in SpaceFM, when you single-click on a file it will update the status bar.  If you double-click it will update the status bar AND THEN open the file double-clicked.  So both clicks are handled.

A script would work the same.  eg a first click on a list item might change some other value in the dialog to reflect which item is selected.  A double-click will choose that item and automatically click OK.  Yet a double-click will first do the value update caused by the single-click.

> I assume I need to handle the event by parsing of '%v' in my script, perhaps something like..

That's the basic idea.  Note that %v will contain quoted elements.  For example, with the --chooser, "%v" will equal something like: "chooser1 '/etc/fstab' '/etc/mtab'"  Note that each value is single-quoted for bash.  Then the whole value is passed as one argument on the command line.

If you run the click command like this:

    --click bash -c 'echo "%v" >&2'

Then since you quoted "%v", it will all be in $1.  That why your script needed the ${1%% *} (which should be "${1%% *}").  However, it's better not to quote %v:

    --click bash -c 'echo %v >&2'

This way, if you run a script instead of echo, $1 will be the element name.  Then you can use 'shift' to get run of $1.  Then the values will be $2, $3, etc.  ("$@").  For example:

    --click bash -c 'myhandler.sh %v'

So myhandler.sh will receive $1=name, $2=value1, $3=value2, etc.  However, without bash:

    --click myhandler.sh %v

Now it will all be in $1.

You can try this script (md5.sh):
~~~~
#!/bin/bash

echo "NAME: $1" >&2
name="$1"
shift

if [ "$name" == "chooser1" ]; then
    md5sum "$@" >&2
fi
~~~~

With this command:

    spacefm -g --chooser --multi --label Nothing \
               --click bash -c 'md5.sh %v'

Note that you can select multiple files (ctrl+click) and they will all be summed everytime you click.

So basically SpaceFM replaces %v with a single command-line argument to the command being run.  If that command is bash, it will then expand that argument into multiple arguments, and will handle filenames containing spaces and other quoted characters correctly.  Maybe I'll add a note on this to the manual.

**Comment by IgnorantGuru** [comment]
Also, if you add multiple --click elements, each one will run.  Not sure why you would do so, but it doesn't prevent you.

**Comment by paoloschi** [comment]
illuminating explanation, now I've really understand everything.
I have no further questions/requests to be placed on this subject.

this is an improvement that will take me to update most of my earlier creations so a lot of work waiting for me..
wonderful! ;-)
